### ISSUE TITLE
Several queries fail with work-stealing issues in 2020-10-20 nightlies

### ISSUE THREAD

**Issue by beckernick** [rapidsai/tpcx-bb#121]
At some point between 2020-10-15 and 2020-10-20 nightlies, we introduced a change in one of the upstream libraries that causes work-stealing issues in some TPCxBB queries. 

I've verified this can happen in several queries, but its most clearly reproduced in Q21. I've copied part of a worker log from Q21 SF1K running on a TCP Dask cluster with 16 GPUs. We don't have any of these issues in an environment from 2020-10-15. In that environment, the worker log is essentially empty minus a few garbage collection statements.

In the 2020-10-20 environment:

```python-traceback
...
distributed.core - ERROR - "('group-shuffle-0-e7b6c92c8cff5e439539bca58cbf8d8d', (4, 5))"
Traceback (most recent call last):
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/core.py", line 597, in handle_stream
    handler(**merge(extra, msg))
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/worker.py", line 2194, in steal_request
    state = self.tasks[key].state
KeyError: "('group-shuffle-0-e7b6c92c8cff5e439539bca58cbf8d8d', (4, 5))"
distributed.worker - ERROR - "('group-shuffle-0-e7b6c92c8cff5e439539bca58cbf8d8d', (4, 5))"
Traceback (most recent call last):
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/worker.py", line 942, in handle_scheduler
    comm, every_cycle=[self.ensure_communicating, self.ensure_computing]
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/core.py", line 597, in handle_stream
    handler(**merge(extra, msg))
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/worker.py", line 2194, in steal_request
    state = self.tasks[key].state
KeyError: "('group-shuffle-0-e7b6c92c8cff5e439539bca58cbf8d8d', (4, 5))"
distributed.worker - INFO - Connection to scheduler broken.  Reconnecting...
distributed.core - ERROR - "('group-shuffle-0-e7b6c92c8cff5e439539bca58cbf8d8d', (1, 1))"
Traceback (most recent call last):
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/core.py", line 597, in handle_stream
    handler(**merge(extra, msg))
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/worker.py", line 2194, in steal_request
    state = self.tasks[key].state
KeyError: "('group-shuffle-0-e7b6c92c8cff5e439539bca58cbf8d8d', (1, 1))"
tornado.application - ERROR - Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOLoop object at 0x7f1f8110e950>>, <Task finished coro=<Worker.handle_scheduler() done, defined at /raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/worker.py:939> exception=KeyError("('group-shuffle-0-e7b6c92c8cff5e439539bca58cbf8d8d', (4, 5))")>)
Traceback (most recent call last):
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/tornado/ioloop.py", line 743, in _run_callback
    ret = callback()
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/tornado/ioloop.py", line 767, in _discard_future_result
    future.result()
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/worker.py", line 942, in handle_scheduler
    comm, every_cycle=[self.ensure_communicating, self.ensure_computing]
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/core.py", line 597, in handle_stream
    handler(**merge(extra, msg))
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/worker.py", line 2194, in steal_request
    state = self.tasks[key].state
KeyError: "('group-shuffle-0-e7b6c92c8cff5e439539bca58cbf8d8d', (4, 5))"
distributed.worker - ERROR - "('group-shuffle-0-e7b6c92c8cff5e439539bca58cbf8d8d', (1, 1))"
Traceback (most recent call last):
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/worker.py", line 942, in handle_scheduler
    comm, every_cycle=[self.ensure_communicating, self.ensure_computing]
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/core.py", line 597, in handle_stream
    handler(**merge(extra, msg))
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/worker.py", line 2194, in steal_request
    state = self.tasks[key].state
KeyError: "('group-shuffle-0-e7b6c92c8cff5e439539bca58cbf8d8d', (1, 1))"
distributed.worker - INFO - Connection to scheduler broken.  Reconnecting...
distributed.worker - INFO - Comm closed
tornado.application - ERROR - Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOLoop object at 0x7f9f30dfcb10>>, <Task finished coro=<Worker.handle_scheduler() done, defined at /raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/worker.py:939> exception=KeyError("('group-shuffle-0-e7b6c92c8cff5e439539bca58cbf8d8d', (1, 1))")>)
Traceback (most recent call last):
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/tornado/ioloop.py", line 743, in _run_callback
    ret = callback()
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/tornado/ioloop.py", line 767, in _discard_future_result
    future.result()
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/worker.py", line 942, in handle_scheduler
    comm, every_cycle=[self.ensure_communicating, self.ensure_computing]
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/core.py", line 597, in handle_stream
    handler(**merge(extra, msg))
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/worker.py", line 2194, in steal_request
    state = self.tasks[key].state
KeyError: "('group-shuffle-0-e7b6c92c8cff5e439539bca58cbf8d8d', (1, 1))"
distributed.worker - INFO - Comm closed
distributed.batched - INFO - Batched Comm Closed:
distributed.batched - INFO - Batched Comm Closed:
distributed.utils_perf - INFO - full garbage collection released 218.97 MB from 166 reference cycles (threshold: 10.00 MB)
distributed.worker - INFO - Comm closed
distributed.worker - INFO - Comm closed
distributed.worker - INFO - Comm closed
distributed.worker - INFO - Comm closed
distributed.worker - INFO - Comm closed
distributed.worker - INFO - Comm closed
distributed.worker - INFO - Comm closed
distributed.worker - INFO - Comm closed
distributed.worker - INFO - Comm closed
distributed.worker - INFO - Comm closed
distributed.worker - INFO - Comm closed
distributed.worker - INFO - Comm closed
distributed.worker - INFO - Comm closed
distributed.worker - INFO - Comm closed
distributed.worker - INFO - Comm closed
distributed.worker - INFO - Comm closed
distributed.utils_perf - INFO - full garbage collection released 110.92 MB from 95 reference cycles (threshold: 10.00 MB)
distributed.worker - ERROR -
Traceback (most recent call last):
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/worker.py", line 1573, in transition_flight_memory
    self.batched_stream.send({"op": "add-keys", "keys": [ts.key]})
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/batched.py", line 146, in send
    raise CommClosedError
distributed.comm.core.CommClosedError
tornado.application - ERROR - Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOLoop object at 0x7f1f8110e950>>, <Task finished coro=<Worker.gather_dep() done, defined at /raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/worker.py:1960> exception=CommClosedError()>)
Traceback (most recent call last):
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/tornado/ioloop.py", line 743, in _run_callback
    ret = callback()
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/tornado/ioloop.py", line 767, in _discard_future_result
    future.result()
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/worker.py", line 2073, in gather_dep
    self.transition(ts, "memory", value=data[d])
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/worker.py", line 1499, in transition
    state = func(ts, **kwargs)
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/worker.py", line 1573, in transition_flight_memory
    self.batched_stream.send({"op": "add-keys", "keys": [ts.key]})
  File "/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020/lib/python3.7/site-packages/distributed/batched.py", line 146, in send
    raise CommClosedError
distributed.comm.core.CommClosedError
...
```

Environments:

2020-10-15 (works fine)
```
conda list | grep "rapids\|dask\|distr"
# packages in environment at /raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201015:
cudf                      0.16.0a201015   cuda_10.2_py37_gc6dfa6e3ed_1977    rapidsai-nightly
cuml                      0.16.0a201015   cuda10.2_py37_gb869880a3_867    rapidsai-nightly
dask                      2.30.0+12.g14f6a76b          pypi_0    pypi
dask-cuda                 0.16.0a201015           py37_96    rapidsai-nightly
dask-cudf                 0.16.0a201015   py37_gc6dfa6e3ed_1977    rapidsai-nightly
distributed               2.30.0+6.g2b43b402          pypi_0    pypi
faiss-proc                1.0.0                      cuda    rapidsai-nightly
libcudf                   0.16.0a201015   cuda10.2_gc6dfa6e3ed_1977    rapidsai-nightly
libcuml                   0.16.0a201015   cuda10.2_gb869880a3_867    rapidsai-nightly
libcumlprims              0.16.0a200930   cuda10.2_g1c28023_35    rapidsai-nightly
librmm                    0.16.0a201015   cuda10.2_gb7cdbc4_399    rapidsai-nightly
rmm                       0.16.0a201015   cuda_10.2_py37_gb7cdbc4_399    rapidsai-nightly
ucx                       1.8.1+g6b29558       cuda10.2_0    rapidsai-nightly
ucx-proc                  1.0.0                       gpu    rapidsai-nightly
ucx-py                    0.16.0a201015   py37_g6b29558_187    rapidsai-nightly
```

2020-10-20 (failing)
```
conda list | grep "rapids\|dask\|distr"
# packages in environment at /raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20201020:
cudf                      0.16.0a201020   cuda_10.2_py37_ga6296e34b8_1978    rapidsai-nightly
cuml                      0.16.0a201017   cuda10.2_py37_g48a5416c8_873    rapidsai-nightly
dask                      2.30.0+22.g2b45fdb9          pypi_0    pypi
dask-cuda                 0.16.0a201020           py37_96    rapidsai-nightly
dask-cudf                 0.16.0a201020   py37_ga6296e34b8_1978    rapidsai-nightly
distributed               2.30.0+8.ga0b2a605          pypi_0    pypi
faiss-proc                1.0.0                      cuda    rapidsai-nightly
libcudf                   0.16.0a201020   cuda10.2_ga6296e34b8_1978    rapidsai-nightly
libcuml                   0.16.0a201017   cuda10.2_g48a5416c8_873    rapidsai-nightly
libcumlprims              0.16.0a200930   cuda10.2_g1c28023_35    rapidsai-nightly
librmm                    0.16.0a201020   cuda10.2_gb7cdbc4_399    rapidsai-nightly
rmm                       0.16.0a201020   cuda_10.2_py37_gb7cdbc4_399    rapidsai-nightly
ucx                       1.8.1+g6b29558       cuda10.2_0    rapidsai-nightly
ucx-proc                  1.0.0                       gpu    rapidsai-nightly
ucx-py                    0.16.0a201020   py37_g6b29558_187    rapidsai-nightly
```

**Comment by beckernick** [comment]
cc @quasiben @kkraus14

**Comment by kkraus14** [comment]
I suspect it's an upstream Dask / Distributed change. Any chance you can try older versions of those with up to date RAPIDS libraries to see if it reproduces?

**Comment by quasiben** [comment]
Does this happen at the beginning of the workflow ?  cc @madsbk in case he thoughts about recent HLG changes

**Comment by beckernick** [comment]
Can confirm that 2020-10-20 rapids nightlies + dask and distributed at the following commits (same as in the Oct 15th environment) does not have the work-stealing errors.

Dask at 14f6a76b38b2990fa40ef64fcd8933b6ee56b93f
Distributed at 2b43b402c19b278a1e2fe7e1bf7718452b2fe007

**Comment by beckernick** [comment]
> Does this happen at the beginning of the workflow ? cc @madsbk in case he thoughts about recent HLG changes

Not immediately. Some work happens and then we appear to hang after some work-stealing errors.

**Comment by beckernick** [comment]
@quasiben can confirm this commit in distributed is the breaking one https://github.com/dask/distributed/commit/dcb46d07318282d8a8976aadb0e4003144a42ae7

**Comment by jakirkham** [comment]
It's interesting that the key is `('group-shuffle-0-e7b6c92c8cff5e439539bca58cbf8d8d', (4, 5))`. Historically we have done something more like `('group-shuffle-0-e7b6c92c8cff5e439539bca58cbf8d8d', 4, 5)` (or at least this is the case in other Dask operations)

cc @rjzamora

**Comment by quasiben** [comment]
@beckernick and I went through this with latest dask and things seem to be working though we do see errors like the following in worker logs:


```
distributed.utils - ERROR - "('split-shuffle-1-a0144b78484e8a3c85a6cf4105a8efc9', 5, (6, 7))"
Traceback (most recent call last):
  File "/datasets/bzaitlen/GitRepos/distributed/distributed/utils.py", line 655, in log_errors
    yield
  File "/datasets/bzaitlen/GitRepos/distributed/distributed/worker.py", line 1989, in gather_dep
    deps_ts = [self.tasks[key] for key in deps]
  File "/datasets/bzaitlen/GitRepos/distributed/distributed/worker.py", line 1989, in <listcomp>
    deps_ts = [self.tasks[key] for key in deps]
KeyError: "('split-shuffle-1-a0144b78484e8a3c85a6cf4105a8efc9', 5, (6, 7))"
```

While these don't prevent successful runs they are a bit concerning.  At the moment, I think these are due to changes in the scheduler code and should be resolved as work continues improving the scheduler

**Comment by quasiben** [comment]
@beckernick I think this issue is now being tracked in distributed: https://github.com/dask/distributed/issues/4231 Once Gil's has a PR in we should test this again.

**Comment by beckernick** [comment]
Excellent! Thanks for cross-linking @quasiben and thanks @madsbk for creating a minimal reproducer 😄

**Comment by gforsyth** [comment]
Fix merged upstream (dask/distributed#4241) so y'all should be good to restart those benchmarks.

**Comment by beckernick** [comment]
This is now resolved with dask and distributed master. Thanks for the upstream fix @gforsyth !

Will file an issue to unpin or update the dask and distributed commits in this repo's conda spec file.
